srml/contract/src/account_db.rs

```diff
@@ -131,18 +83,42 @@ impl<T: Trait> AccountDb<T> for DirectAccountDb {
 					continue;
 				}
 			}
-			if let Some(code) = changed.code {
-				if let Some(code) = code {
-					<CodeHashOf<T>>::insert(&address, code);
+			if changed.code.is_some() || !changed.storage.is_empty() {
+				let mut info = if !<AccountInfoOf<T>>::exists(&address) {
+					let info = AccountInfo {
+						trie_id: <T as Trait>::TrieIdGenerator::trie_id(&address),
+						current_mem_stored: 0,
+					};
+					<AccountInfoOf<T>>::insert(&address, &info);
+					info
 				} else {
-					<CodeHashOf<T>>::remove(&address);
+					<AccountInfoOf<T>>::get(&address).unwrap()
+				};
+
+				if let Some(code) = changed.code {
+					if let Some(code) = code {
+						<CodeHashOf<T>>::insert(&address, code);
+					} else {
+						<CodeHashOf<T>>::remove(&address);
+					}
 				}
-			}
-			for (k, v) in changed.storage.into_iter() {
-				if let Some(value) = v {
-					child::put_raw(&trieid[..], &k, &value[..]);
-				} else {
-					child::kill(&trieid[..], &k);
+
+				let mut new_mem_stored = info.current_mem_stored;
+				for (k, v) in changed.storage.into_iter() {
+					if let Some(value) = child::get::<Vec<u8>>(&info.trie_id[..], &k) {
+						new_mem_stored -= value.len() as u64;
+					}
+					if let Some(value) = v {
+						new_mem_stored += value.len() as u64;
+						child::put_raw(&info.trie_id[..], &k, &value[..]);
+					} else {
+						child::kill(&info.trie_id[..], &k);
+					}
+				}
+
+				if new_mem_stored != info.current_mem_stored {
+					info.current_mem_stored = new_mem_stored;
```

Comment:
I'm up for renaming this from `mem` to `storage`, since `mem` has a slight associations with a RAM.